ek9k status and configuration records

JJL772 opened the pull request:
These records were in the module before, but it needed some improvement. The main big change here is reading all status records in one go every ~1 second, rather than having each individually poll. Additionally, alarms will be set accordingly if status records could not be read, which should help us keep track of which devices are unresponsive and when they entered that state.

In the latest commit, some of the alarm and error handling has been cleaned up. Instead of setting `READ_ALARM`/`WRITE_ALARM` when I/O fails, it sets `COMM_ALARM`. For terminals with limits built-in, `HW_LIMIT_ALARM` is set when those are exceeded. This component probably needs some more in-depth review, I haven't been able to find any documentation on `COMM_ALARM` or the other alarm types except `READ_ALARM` and `WRITE_ALARM`.

I added a couple status/configuration screen as well, it looks like this:
![Screenshot_20230630_165753](https://github.com/slaclab/epics-ek9000/assets/19717056/277d2014-0197-44bc-a150-f50ddf8ac943)
Everything in the coupler state section is read-only.

JJL772 committed:
FIX: ek9000 coupler records

JJL772 committed:
ENH: add coupler settings screen

JJL772 committed:
ENH: improve ek9000 status and configuration records

JJL772 committed:
STY: apply clang-format

JJL772 committed:
FIX: warning fixes and couple screen title

JJL772 committed:
ENH: cleanup alarm logic

Use COMM_ALARM and HW_LIMIT_ALARM where acceptable instead of READ_ALAM/WRITE_ALARM

klauer commented on ek9000App/src/devEL1XXX.cpp:
```diff
@@ -124,19 +121,13 @@ template <class RecordT> static long EL10XX_read_record(void* prec) {
-		LOG_WARNING(dpvt->pdrv, "EL10XX_read_record() for %s: %s\n", pRecord->name, devEK9000::ErrorToString(status));
-		return 0;
+	if (status != EK_EOK) {
+		recGblSetSevr(pRecord, COMM_ALARM, INVALID_ALARM);
```
On the client side, we care most about the alarm severity which is set appropriately here as `INVALID`. 

It's possible others may disagree, but in my opinion the alarm status of "READ" vs "COMM" is just useful metadata for the user to differentiate between different potential scenarios. "Communication with the controller failed" vs "my attempt to read data failed (for some other reason)" is how I'd interpret them.

(edit: I meant to indicate that I agreed with your change and that `COMM_ALARM` is just fine)

JJL772 replied:
Thanks for the input! I was unsure of this change and was totally unaware of the other alarm types until last week when I was reading some random device support code.

klauer commented on ek9000App/src/devEK9000.h:
```diff
@@ -82,6 +86,13 @@ enum {
 };
 
+/* Buffered IO types */
+enum EIOType {
```
`int` -> sensible enum is always a 👍 from me

klauer commented on ek9000App/src/devEL3XXX.cpp:
```diff
@@ -333,36 +325,27 @@ static long EL36XX_read_record(void* prec) {
-	}
-
-	/* Set props */
-	pRecord->pact = FALSE;
```
Just checking - `pact = FALSE` removal was intentional?

JJL772 replied:
My understanding is that `PACT` isn't set to `true` by EPICS implicitly, and is intended to be used with callbacks. Maybe I misunderstood?

klauer replied:
Yep, that's right

JJL772 replied:
The app developer guide seems to confirm this: 
![image](https://github.com/slaclab/epics-ek9000/assets/19717056/ff2c5702-d1f8-415b-91a0-5e9b4ced37a6)

https://epics.anl.gov/base/R3-16/2-docs/AppDevGuide.pdf pg 196-197 has the example synchronous device support module.

I believe the `PACT` here was leftover from when this was using callbacks instead of buffered IO from the device.

klauer commented on ek9000App/src/ekUtil.h:
```diff
@@ -251,6 +256,11 @@ template <> inline bool setupCommonDpvt<aoRecord>(aoRecord* prec, TerminalDpvt_t
 }
 
+inline bool parseInt(const std::string& s, int& out) {
+	bool hex = s[0] == '0' && s[1] == 'x';
```
Check for `s.length() > 2`?

Er, wait, where did this function sneak in? Doesn't seem used anywhere...

JJL772 replied:
Wow I wrote a useful function and didn't even use it...

klauer commented on ek9000App/src/devEL50XX.cpp:
```diff
@@ -144,21 +144,22 @@ static long el50xx_read_record(void* prec) {
-		return 0;
-	}
+	if (!util::DpvtValid(dpvt))
+		return 1;
```
So before, each of these checked their connectivity by way of `VerifyConnection()`, but now we rely on the EK9K to track its status and let us know about a bad state by way of `getEK9000IO()` and from there we set our `COMM_ALARM`/`INVALID` severity. 

Makes sense 👍

JJL772 replied:
Yep, `VerifyConnection()` is only useful when we want to avoid calling `doModbusIO` on a disconnected device, as that results in a long timeout. The branch here was leftover from when this was using callbacks instead of buffered IO.

klauer commented on ek9000App/src/devEK9000.cpp:
```diff
@@ -1457,167 +1494,231 @@ int CoE_ParseString(const char* str, ek9k_coe_param_t* param) {
+};
+
+CONSTEXPR StatusReg status_regs[] = {{"analogOutputs", 0x1010, STATUS_RD | STATUS_STATIC},
+									 {"analogInputs", 0x1011, STATUS_RD | STATUS_STATIC},
```
Having these by name is convenient for sure.  👍 

Are there any reserved values in this status register block that the user may end up wanting to access down the line? If you think there's potential for it, it may be beneficial to offer the ability to specify the status register address/flags in the instio string itself so they don't have to rewrite the EK9K code to allow for it.

JJL772 replied:
Probably a good idea. This does cover virtually all of the status registers though, except 0x6000-0x60FF, which contain rail layout.

klauer commented on ek9000App/src/devEK9000.cpp:
```diff
@@ -1457,167 +1494,231 @@ int CoE_ParseString(const char* str, ek9k_coe_param_t* param) {
+				}
+			}
+			if (param.reg == 0) {
+				epicsPrintf("Malformed integer value '%s' in instio string '%s'\n", spec[i].second.c_str(), str);
```
```suggestion
				epicsPrintf("Malformed instio string '%s' does not specify status register\n", str);
```

Maybe?

klauer reviewed (commented):
I think this looks good. 

There might be some overlap with #47 and since that one's smaller perhaps it should go in first?

JJL772 committed:
ENH: allow arbitrary registers to be read out of the coupler's address space

JJL772 commented:
Ok, rebased on master and applied the suggestions.

JJL772 committed:
FIX: disable -Wshadow for now

The static assert macro can cause shadowed variables, apparently. Will need to fix this eventually

klauer reviewed (approved):
Changes since my last review look good
I'd say this is ready 👍

JJL772 commented:
Thanks for the review!
